### PR TITLE
Seun - Synapses Part 1

### PR DESCRIPTION
**Pull request by ogewan**
Annotation overhaul

Summary of Changes:

- New annotations
    - Collection (Meta-Annotation)
        - Generic collection of any annotation including LineStrip and Collection
    - LineStrip (Multiple connected point annotation), subclass of Collection
        - Immutable dynamic size annotation
        - Supports looping, last annotation connects to the first
- Arbitrary step annotation construction
    - Confirm Annotation - Completes a multistep annotation
    - Abort Annotation - Cancels a multistep annotation, deleting it and its children
- Contextual Annotation Controls
    - Multiple select - Can select multiple annotations at once
    - Delete Annotation - Deleting a collection deletes its child annotation
    - Create Collection - Form a collection with the selected annotation(s) as children. If the first annotation is a child annotation, the new collection will become a child of the first annotation's old parent.
    - Free Annotations - Remove all children from a collection and delete the now empty parent. If the parent is a child annotation, the freed annotation's become children of the parent's parent.
    - Remove from Collection - Remove a single child from a collection, deleting the collection if it becomes empty. If the parent has a parent, this will become the child's new parent. Note: LineStrips are immutable, so its children cannot be removed like this.
- Export CSV
    - Supports Collection and LineStrip
    - Change in format to allow for arbitrary size collections
- Import CSV
    - [x] Transfer to separate branch?

### CODE REVIEW

**Comment by manuel-castro** [comment]
44 commits, 16 files changed, 2500 line diff :skull:

File: src/neuroglancer/annotation/collection.ts [hunk 1]
```diff
@@ -0,0 +1,309 @@
...
+    return annotation;
+  }
+});
+
+export type MultiTool =
+    typeof PlacePointTool|typeof PlaceBoundingBoxTool|typeof PlaceLineTool|typeof PlaceSphereTool|
+    typeof PlaceLineStripTool|typeof PlaceSpokeTool|typeof MultiStepAnnotationTool;
+
+
+export class MultiStepAnnotationTool extends PlaceAnnotationTool {
```
Comment (manuel-castro) [inline src/neuroglancer/annotation/collection.ts]: I think we should add an additional abstract class that extends the PlaceAnnotationTool that MultiStepAnnotationTool, LineStrip, and Spoke all inherit from. That way we can avoid having a class that has code that is never meant to be executed for instances of that class.

I also think we should add some access level modifiers to the methods of this class. E.g. updateLast and getChildRef are only called within the class so it should be private; appendNewChildAnnotation is called by children so it should be protected.

File: src/neuroglancer/annotation/annotation.ts [hunk 2]
```diff
@@ -0,0 +1,168 @@
+import {Annotation, AnnotationReference, AnnotationType, AxisAlignedBoundingBox, Collection, Line} from 'neuroglancer/annotation';
+import {AnnotationLayerState} from 'neuroglancer/annotation/frontend';
+import {MouseSelectionState} from 'neuroglancer/layer';
+import {StatusMessage} from 'neuroglancer/status';
+import {UserLayerWithAnnotations} from 'neuroglancer/ui/annotations';
+import {Tool} from 'neuroglancer/ui/tool';
+import {vec3} from 'neuroglancer/util/geom';
+import {verifyObjectProperty, verifyOptionalString} from 'neuroglancer/util/json';
+import {Uint64} from 'neuroglancer/util/uint64';
+import {MultiStepAnnotationTool} from './collection';
```
Comment (manuel-castro) [inline src/neuroglancer/annotation/annotation.ts]: Formatting

File: package-lock.json [hunk 3]
```diff
@@ -3213,8 +3213,7 @@
         },
         "ansi-regex": {
           "version": "2.1.1",
-          "bundled": true,
-          "optional": true
```
[the remark this answers is not in the document]
Comment (ogewan) [reply package-lock.json]: I'm running on npm 6.11.1

File: src/neuroglancer/ui/annotations.ts [hunk 4]
```diff
@@ -641,26 +869,36 @@ export class AnnotationLayerView extends Tab {
   }
 
   private updateAnnotationElement(annotation: Annotation, checkVisibility = true) {
+    // TODO:
+    // https://github.com/google/neuroglancer/commit/a05650cd604887dd6ba8a23855a7afe99a89b7d0#r35414033
+    // TODO: Rebase to master and introduce jeremy's changes
```
[the remark this answers is not in the document]
Comment (ogewan) [reply src/neuroglancer/ui/annotations.ts]: Need to merge in Jeremy's fix and rebase to that

**Comment by ogewan** [comment]
relocated to [seun-annotations-pt1](https://github.com/seung-lab/neuroglancer/tree/seun-annotations-pt1)

**Pull request closed without merging**